src/mono/mono/metadata/gc.c

```diff
@@ -694,8 +694,8 @@ mono_gc_finalize_notify (void)
 	if (mono_gc_is_null ())
 		return;
 
-#if defined(HOST_WASI)
-	// TODO: Schedule the background job on WASI. Threads aren't yet supported in this build.
+#if defined(HOST_WASI) && defined(DISABLE_THREADS)
+	mono_runtime_do_background_work ();
```

Comment:
I'm not confident what are problems of doing it here synchronously and inline, for single-threaded runtime. Will test and see what happens.